### PR TITLE
[ENH] code sktime/forecasting/gluonts_adapters.py

### LINKED ISSUE #6800: [ENH] `gluonts` models
It would be nice to have `gluonts` models interface and indexed.

These are mainly a handful of torch/lightning based models with a consistent, `torch` based interface.

There is recent work towards adding the `gluonts` data containers as mtypes, #6530, which should allow to add these easily.

Contingent on 6530 being merged - good first issue in collaboration with current workstreams.

### PR DESCRIPTION
**Pull request by rajatkriplani**
#### Reference Issues/PRs
Fixes #6800 
#### What does this implement/fix? 
This PR introduces modular improvements and refactoring for the `gluonts_adapters.py` module in `sktime`. It addresses repetitive method overrides, improves modularity, and integrates the newly implemented conversion functions.  
Key changes include:
- **Refactored Base Adapter**:
  - Centralized data conversion logic into reusable methods (`_convert_to_gluonts_format`, `_convert_to_sktime_format`).
  - Modularized model initialization by introducing `_init_model` and enabling dynamic model class specification.
- **DeepAR and N-BEATS Adapters**:
  - Simplified implementation by leveraging the refactored base adapter.
  - Eliminated duplication of trainer initialization and prediction logic.
- **Integration of New Conversion Utilities**:
  - Used `convert_gluonts_result_to_multiindex` and `convert_from_multiindex_to_listdataset` for improved compatibility with `sktime` data formats.

These changes streamline the codebase, reduce redundancy, and make it easier to add new GluonTS-based models in the future.
#### Does your contribution introduce a new dependency? If yes, which one?
No
#### What should a reviewer concentrate their feedback on?
Please focus on:
- The correctness and clarity of the refactored modular design.
- Compatibility with existing `sktime` functionalities and GluonTS models.
- Adequacy of the integration of new conversion functions.
- Potential edge cases or scenarios not covered in the current implementation.

#### Did you add any tests for the change?
No
#### Any other comments?
No
##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned.
Note: I can't locate `blob/main/.all-contributorsrc`
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by rajatkriplani** [comment]
Hi @fkiraly I'm fixing the code for code-quality checks but I'm facing an issue. I imported the following:
```python
from sktime.datatypes._adapter.gluonts_to_pd_multiindex import convert_gluonts_result_to_multiindex
from sktime.datatypes._adapter.pd_multiindex_to_list_dataset import convert_from_multiindex_to_listdataset
from sktime.forecasting.base._base import BaseForecaster
```
but these are not being imported. I tried setting up `PYTHONPATH` by:
```bash
export PYTHONPATH=/home/divya/Code/open_scr:$PYTHONPATH
source ~/.bashrc
```

**Comment by rajatkriplani** [comment]
@fkiraly I'm making changes in the code for code-quality checks for that I'm trying to run/test tings locally but I'm unable to import the follwoing:
```python
from sktime.datatypes._adapter.gluonts_to_pd_multiindex import convert_gluonts_result_to_multiindex
from sktime.datatypes._adapter.pd_multiindex_to_list_dataset import convert_from_multiindex_to_listdataset
from sktime.forecasting.base._base import BaseForecaster
```
I even tried setting up the `PYTHONPATH` manually by:
```bash
export PYTHONPATH=[path to sktime repo on my pc]:$PYTHONPATH
source ~/.bashrc
```

note: I've all the requirement fulfilled on the conda environment

**Comment by yarnabrina** [comment]
Hi @rajatkriplani , was the issue solved? If not, can you please give a bit more details? For example what was the exact error, what was traceback message, do you have editable install, etc.?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Tests are failing due to code formatting, see the guide here:
https://www.sktime.net/en/latest/developer_guide/coding_standards.html

**Comment by rajatkriplani** [comment]
Hi @yarnabrina sorry for not addressing your comment earlier, I'll ping you in the thread about this issue over discord. And @fkiraly  I've changed the code format, but I'm facing some import issue on my local repo:

Screenshot from 2025-01-04 01-26-58
I've made sure to import everything correctly, and uninstalled and re-installed all the required dependencies (dev installation), still facing this issue. Can you tell me what else I can do to fix it and test the code?
below is the test code (yet to commit):
```python
import pytest
import pandas as pd
import numpy as np
from sktime.utils.estimator_checks import check_estimator
from sktime.forecasting.base import ForecastingHorizon
from gluonts_adapters import DeepARAdapter, NBEATSAdapter

# 1. Interface Validation with `check_estimator`
def test_check_estimator():
    """Test interface compliance for DeepARAdapter and NBEATSAdapter."""
    check_estimator(DeepARAdapter)
    check_estimator(NBEATSAdapter)

# 2. Unit Tests for `_fit` and `_predict`
def test_fit_and_predict():
    """Test the _fit and _predict methods."""
    y = pd.Series(np.random.rand(100), index=pd.date_range("2023-01-01", periods=100))
    fh = ForecastingHorizon(np.arange(1, 6), is_relative=True)

    # Test DeepARAdapter
    forecaster = DeepARAdapter(prediction_length=5, epochs=1)
    forecaster.fit(y)
    assert forecaster._is_fitted, "DeepARAdapter is not fitted after calling fit."
    y_pred = forecaster.predict(fh)
    assert isinstance(y_pred, pd.DataFrame), "DeepARAdapter prediction output is not a DataFrame."

    # Test NBEATSAdapter
    forecaster = NBEATSAdapter(prediction_length=5, epochs=1)
    forecaster.fit(y)
    assert forecaster._is_fitted, "NBEATSAdapter is not fitted after calling fit."
    y_pred = forecaster.predict(fh)
    assert isinstance(y_pred, pd.DataFrame), "NBEATSAdapter prediction output is not a DataFrame."
```

File: sktime/forecasting/gluonts_adapters.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    convert_gluonts_result_to_multiindex,
+)
+from sktime.datatypes._adapter.pd_multiindex_to_list_dataset import (
+    convert_from_multiindex_to_listdataset,
+)
+
+from sktime.forecasting.base._base import BaseForecaster
+
+
+class GluonTSBaseAdapter(BaseForecaster):
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:18]: Is this class intended to be public to sktime users? If not, I suggest you prefix this with `_` and create within `sktime/forecasting/base/adapters/_gluonts.py` or similarly named module, and then inherit these for concrete estimators in `sktime/forecasting/gluonts.py` or a similarly named module.

If it was intended to be public, then it probably be placed under `sktime/forecasting/adapters`, but @fkiraly can comment on that.

File: sktime/forecasting/gluonts_adapters.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """Common adapter base class for GluonTS models."""
+
+    def __init__(self, model_class, freq="D", prediction_length=24, epochs=10, **kwargs):
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:21]: While I'm personally fi e with this format, I believe it's inconsistent with sktime codebase in general and also with sklearn estimator guidelines. The guidelines suggest to not have any unspecified argument in `__init__` methods, and allowing arbitrary `kwargs` break that. See this: https://scikit-learn.org/stable/developers/develop.html#instantiation

(It's probably not the best link, but that's all I could find now.)

File: sktime/forecasting/gluonts_adapters.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        """Update or retrain the model with new data."""
+        gluonts_data = self._convert_to_gluonts_format(y_new)
+        self.model.train(gluonts_data)
+        return self
+
+
+class DeepARAdapter(GluonTSBaseAdapter):
+    """Adapter for the GluonTS DeepAR model."""
+
+    def __init__(self, **kwargs):
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:85]: Affected by the kwargs comment

File: sktime/forecasting/gluonts_adapters.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        from gluonts.torch.model.deepar import DeepAREstimator
+        super().__init__(model_class=DeepAREstimator, **kwargs)
+
+
+class NBEATSAdapter(GluonTSBaseAdapter):
+    """Adapter for the GluonTS N-BEATS model."""
+
+    def __init__(self, **kwargs):
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:93]: Affected by the kwargs comment

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I've made a few comments, kindly take a look. Also, code quality checks are failing, please ensure you are using pre-commit as documented in the contribution guidelines.

The import errors you mentioned is difficult/impossible for me atleast to debug without live checking, so your best bet will probably be to attend one of the Friday connects and do a screen share.

**Comment by rajatkriplani** [comment]
Hi @yarnabrina  I made changes according to your feedback 
 - `GluonTSBaseAdapter` renamed to `_GluonTSBaseAdapter`
 - removed  `**kwargs` parameter
 - added `trainer_kwargs` instead (additional parameters for the `Trainer`)
 - docstrings added to each `__init__`
 - classes explicitly define parameters
 - `_is_fitted` flag sets True after training (in `_fit`)

**Comment by yarnabrina** [comment]
@rajatkriplani please excuse me if I am missing something obvious, but I am unable to see a single thing you mentioned above. Is there a chance that you not have committed or committed but not pushed?

**Comment by rajatkriplani** [comment]
> @rajatkriplani please excuse me if I am missing something obvious, but I am unable to see a single thing you mentioned above. Is there a chance that you not have committed or committed but not pushed?

@yarnabrina  Sorry for the inconvenience. I accidently committed the previous version again. Now it's fixed please have a look and let me know your feedback.

File: sktime/forecasting/gluonts_adapters.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+This module includes base and specific adapters to use GluonTS models, like
+DeepAR and N-BEATS, in sktime workflows.
+"""
+
+import pandas as pd
+from typing import Optional, Dict, Any
+
+from sktime.forecasting.base import BaseForecaster
+from gluonts.dataset.common import ListDataset
+from gluonts.mx.trainer import Trainer
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:13]: These are problematic. These imports assume gluonts is already present in user's environment, which may not be the case. And if someone wants to use sktime without using your estimator, they can not as these lines will cause failures.

Please take a look at https://www.sktime.net/en/stable/developer_guide/dependencies.html#handling-soft-dependencies. Essentially you have to remove these from module level imports, and move those to methods of the class where it'll be ensured that method does not run unless the dependencies are installed.

File: sktime/forecasting/gluonts_adapters.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    as well as the initialization and training of GluonTS models.
+    """
+
+    def __init__(
+        self,
+        model_class: Any,
+        freq: str = "D",
+        prediction_length: int = 24,
+        epochs: int = 10,
+        trainer_kwargs: Optional[Dict[str, Any]] = None,
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:36]: Small note about type hint here: we have moved on to python 3.9+ support, and we can use `dict` directly. You need not use `typing.Dict` any longer.

File: sktime/forecasting/gluonts_adapters.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    Common adapter base class for GluonTS models.
+
+    This class serves as a base for integrating GluonTS models with sktime's forecasting framework.
+    It handles the conversion between sktime's data formats and GluonTS's data formats,
+    as well as the initialization and training of GluonTS models.
+    """
+
+    def __init__(
+        self,
+        model_class: Any,
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:32]: I am not sure how I feel about this to be honest. You actually do not support anything, it has to be a estimator class from gluonts. But I am unable to think of a better way to type hint here, so let's wait for review from rest of @sktime/core-developers.

File: sktime/forecasting/gluonts_adapters.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            GluonTS model class to instantiate.
+        freq : str, optional (default="D")
+            Frequency of the time series.
+        prediction_length : int, optional (default=24)
+            Number of time steps to predict.
+        epochs : int, optional (default=10)
+            Number of training epochs.
+        trainer_kwargs : dict, optional (default=None)
+            Additional keyword arguments for the GluonTS Trainer.
+        """
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:53]: It is customary that `__init__` methods do not need docstrings of their own, and they are documented as part of the class docstring itself. You can move this section to the above.

File: sktime/forecasting/gluonts_adapters.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        self.freq: str = freq
+        self.prediction_length: int = prediction_length
+        self.epochs: int = epochs
+        self.model_class: Any = model_class
+        self.trainer_kwargs: Dict[str, Any] = trainer_kwargs or {}
+
+        self.model = self._init_model()
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:60]: My suggestion will be to not call this private method here. Initiate the `self.model` attribute as `None` here, and then call the method during `_fit`.

File: sktime/forecasting/gluonts_adapters.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        self._is_fitted: bool = False
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:61]: Why do you need to define this? It is already part of `BaseForecaster`, any reason to override?

File: sktime/forecasting/gluonts_adapters.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        epochs : int, optional (default=10)
+            Number of training epochs.
+        trainer_kwargs : dict, optional (default=None)
+            Additional keyword arguments for the GluonTS Trainer.
+        """
+        self.freq: str = freq
+        self.prediction_length: int = prediction_length
+        self.epochs: int = epochs
+        self.model_class: Any = model_class
+        self.trainer_kwargs: Dict[str, Any] = trainer_kwargs or {}
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:58]: Please do not do this, as it breaks the sktime convention (or possibly in general sklearn convention) that user passed arguments are available as it is as instance properties. You should keep `self.trainer_kwargs` as `trainer_kwargs` only, but initiate a new private property `self._trainer_kwargs` with the `None` handling.

File: sktime/forecasting/gluonts_adapters.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        ----------
+        y : pd.Series
+            The time series data to convert.
+
+        Returns
+        -------
+        gluonts_dataset : ListDataset
+            The converted GluonTS ListDataset.
+        """
+        return convert_from_multiindex_to_listdataset(y)
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:95]: Why do we need this method, and not call the imported function directly?

File: sktime/forecasting/gluonts_adapters.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+        ----------
+        gluonts_result : Any
+            The forecast result from GluonTS.
+
+        Returns
+        -------
+        sktime_forecast : pd.DataFrame
+            The converted forecast in sktime's MultiIndex format.
+        """
+        return convert_gluonts_result_to_multiindex(gluonts_result)
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:111]: Same confusion as above.

File: sktime/forecasting/gluonts_adapters.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+    def _fit(self, y: pd.Series, X: Optional[Any] = None, fh: Optional[Any] = None) -> "_GluonTSBaseAdapter":
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:113]: If you are into type hints, why skip `self`? :)

File: sktime/forecasting/gluonts_adapters.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+            Forecasting horizon.
+
+        Returns
+        -------
+        self : _GluonTSBaseAdapter
+            The fitted adapter instance.
+        """
+        gluonts_data = self._convert_to_gluonts_format(y)
+        self.model.train(gluonts_data)
+        self._is_fitted = True
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:133]: I'm quite certain that this is unnecessary, as it is taken care by `BaseForecaster`.

File: sktime/forecasting/gluonts_adapters.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+            Optional exogenous variables.
+
+        Returns
+        -------
+        y_pred : pd.DataFrame
+            The predicted values in sktime's MultiIndex format.
+        """
+        gluonts_data = self._convert_to_gluonts_format(self._y)
+        forecast_it, _ = self.model.make_evaluation_predictions(
+            gluonts_data, num_samples=100
```
Comment (yarnabrina) [inline sktime/forecasting/gluonts_adapters.py:154]: Should we let user specify `num_samples` themselves?

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I've made few comments, and these are not super exhaustive. Please take a look, and let me know if it makes sense to you and/or you need any help from us.

Also, please take a look at https://www.sktime.net/en/stable/developer_guide/coding_standards.html.

File: sktime/forecasting/gluonts_adapters.py (hunk 5 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:13]: You're absolutely right, I'll refactor the code. And I'll go through the doc once I have replied to your feedback.

File: sktime/forecasting/gluonts_adapters.py (hunk 8 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:53]: I'll consolidate the parameter descriptions into the class-level docstring.

File: sktime/forecasting/gluonts_adapters.py (hunk 9 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:60]: Working on it along with other stuff.

File: sktime/forecasting/gluonts_adapters.py (hunk 10 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:61]: Yes thanks for points that out, I will make sure to above redundancy in code from now on.

File: sktime/forecasting/gluonts_adapters.py (hunk 12 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:95]: Now that I think of it this is just another failed attempt to solve the import issue. I hope you are available for live check-up tomorrow after the meet-up.

File: sktime/forecasting/gluonts_adapters.py (hunk 13 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:111]: Same import issue.

File: sktime/forecasting/gluonts_adapters.py (hunk 16 above)
Comment (rajatkriplani) [reply sktime/forecasting/gluonts_adapters.py:154]: Oh yes for more flexibility, thanks once again.

**Comment by OmBiradar** [comment]
Can I work on this @rajatkriplani @yarnabrina if no one is currently working on this?
